tiledb/sm/query/writer.cc

```diff
@@ -2197,6 +2198,38 @@ void Writer::optimize_layout_for_1D() {
     layout_ = array_schema_->cell_order();
 }
 
+Status Writer::check_extra_element() {
+  for (const auto& it : buffers_) {
+    const auto& attr = it.first;
+    if (!array_schema_->var_size(attr) || array_schema_->is_dim(attr))
+      continue;
+
+    const void* buffer_off = it.second.buffer_;
+    uint64_t* buffer_off_size = it.second.buffer_size_;
+    const auto num_offsets = *buffer_off_size / constants::cell_var_offset_size;
+    const uint64_t* buffer_val_size = it.second.buffer_var_size_;
+    const uint64_t attr_datatype_size =
+        datatype_size(array_schema_->type(attr));
+    const uint64_t max_offset = offsets_format_mode_ == "bytes" ?
+                                    *buffer_val_size :
+                                    *buffer_val_size / attr_datatype_size;
+    const uint64_t last_offset =
+        get_offset_buffer_element(buffer_off, num_offsets - 1);
+
+    if (last_offset != max_offset)
+      return LOG_STATUS(Status::WriterError(
+          "Invalid offsets for attribute " + attr +
+          "; the last offset: " + std::to_string(last_offset) +
+          " is not equal to the size of the data buffer: " +
+          std::to_string(max_offset)));
+
+    // Remove extra element, it's not needed
+    *buffer_off_size -= constants::cell_var_offset_size;
```

Comment:
The `buffer_off_size` pointer is provided by the user through `Writer::set_buffer`. I don't see anywhere else in the writer where we mutate buffer sizes. This could cause an issue if the user tries to write the same buffer twice, in either serial or parallel.

Since this is for a special case, this is OK. Could you create a ticket to fix this issue in the future? I think the correct fix would be to check for this scenario everytime we access the offsets `buffer_size_` in the writer.

Reply:
you are totally right, let me fix this using story 4612 in clubhouse